activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb

```diff
@@ -557,10 +557,25 @@ def type_to_sql(type, limit: nil, precision: nil, scale: nil, array: nil, **) #
         def columns_for_distinct(columns, orders) #:nodoc:
           order_columns = orders.reject(&:blank?).map { |s|
               # Convert Arel node to string
-              s = s.to_sql unless s.is_a?(String)
+              unless s.is_a?(String)
+                # if this is an order node, then just use the core contents without modification
+                if s.kind_of?(Arel::Nodes::Ordering)
+                  s = s.expr
+                  keep_order = true
+                end
+                # essentially s = s.to_sql but works for attributes too
```

Comment:
This would be better handled using polymorphism by calling `to_sql` and it working for any class. `to_sql` should not be used by apps but rails can use it.